src/xrpld/app/tx/detail/PermissionedDomainSet.cpp

```diff
@@ -0,0 +1,147 @@
+{
+    if (!ctx.rules.enabled(featurePermissionedDomains))
+        return temDISABLED;
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (!ctx.tx.isFieldPresent(sfAcceptedCredentials))
+        return temMALFORMED;
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    // TODO check to see if we should disallow duplicate issuers.
+    // If so, it probably means sorting on the CredentialType field
+    // for identical issuers.
+    if (credentials.empty() || credentials.size() > PD_ARRAY_MAX)
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+XRPAmount
+PermissionedDomainSet::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    if (tx.isFieldPresent(sfDomainID))
+        return Transactor::calculateBaseFee(view, tx);
+    // The fee required for a new PermissionedDomain is one owner reserve.
+    return view.fees().increment;
+}
+
+TER
+PermissionedDomainSet::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount))))
+        return tefINTERNAL;
+
+    assert(ctx.tx.isFieldPresent(sfAcceptedCredentials));
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    for (auto const& credential : credentials)
+    {
+        if (!credential.isFieldPresent(sfIssuer) ||
+            !credential.isFieldPresent(sfCredentialType))
+        {
+            return temMALFORMED;
+        }
+        if (!ctx.view.read(keylet::account(credential.getAccountID(sfIssuer))))
+            return temBAD_ISSUER;
+    }
+
+    if (!ctx.tx.isFieldPresent(sfDomainID))
+        return tesSUCCESS;
+    auto const domain = ctx.tx.getFieldH256(sfDomainID);
+    if (domain == beast::zero)
+        return temMALFORMED;
+    auto const sleDomain = ctx.view.read({ltPERMISSIONED_DOMAIN, domain});
+    if (!sleDomain)
+        return tecNO_ENTRY;
+    auto const owner = sleDomain->getAccountID(sfOwner);
+    auto account = ctx.tx.getAccountID(sfAccount);
+    if (owner != account)
+        return temINVALID_ACCOUNT_ID;
+
+    return tesSUCCESS;
+}
+
+/** Attempt to create the Permissioned Domain. */
+TER
+PermissionedDomainSet::doApply()
+{
+    auto const ownerSle = view().peek(keylet::account(account_));
+
+    // All checks have already been done. Just update credentials. Same logic
+    // for either new domain or updating existing.
+    auto updateSle = [this](std::shared_ptr<STLedgerEntry> const& sle) {
+        auto credentials = ctx_.tx.getFieldArray(sfAcceptedCredentials);
+        // TODO when credentials are merged, it is possible that this will
+        // also need to sort on the CredentialType field in case there
+        // are multiple issuers in each set of credentials. The spec
+        // needs to be ironed out.
+        credentials.sort(
+            [](STObject const& left, STObject const& right) -> bool {
+                return dynamic_cast<STObject const*>(&left)->getAccountID(
```

Comment:
You don't need to cast STObject & to STObject *.

Reply:
fixed